README.md

```diff
@@ -65,55 +85,152 @@ antlr4 -Dlanguage=Python3 -visitor /tests/Expr.g4
 This will generate a number of files in the `/tests/` directory, including a Lexer (`ExprLexer.py`),
 a parser (`ExprParser.py`), and a visitor (`ExprVisitor.py`).
 
-You can use and import these directly in python. For example, from the root of this repo...
+You can use and import these directly in Python. For example, from the root of this repo:
 
 ```bash
 from tests import ExprVisitor
 ```
 
-### Creating AST Classes
+To easily use the generated files, they are put in the `antlr_py` package.
+The `__init__.py` file exports the generated files under an alias that doesn't include the name of the grammar.
+
+### Base nodes
+
+A `BaseNode` subclass has fields for all rule elements and labels for all rule element labels in its corresponding grammar rule.
+Both fields and labels are available as properties on `BaseNode` instances.
+Labels take precedence over fields if the names would collide.
+
+The name of a `BaseNode` is the name of the corresponding ANTLR grammar rule, but starting with an uppercase character.
+If rule alternative labels are specified for an ANTLR rule, these are used instead of the rule name.
+
+### Transforming nodes
+
+#### Transformer
+
+The `BaseNodeTransformer` will walk over the tree from the root node to the leaf nodes.
+When visiting a node, it is possible to transform it.
+The tree is updated with transformed node before continuing the walk over the tree.
+
+To define a node transform, add a static method to the `BaseNodeTransformer` subclass passed to `process_tree`.
+
+- The name of the method you should define follows this pattern: `visit_<BaseNode>`,
+  where `<BaseNode>` should be replaced by the name of the `BaseNode` subclass to transform.
+- The method should return the transformed node.
+
+This is a simple example:
+
+```python
+class Transformer(BaseNodeTransformer):
+    @staticmethod
+    def visit_My_antlr_rule(node):
+        return node.name_of_part
+```
+
+#### Custom nodes
+
+To make it easy to return a custom node, you can define `AliasNode` subclasses.
```

Comment:
Again, might be useful to add an example of when I'd want to use this.

Reply:
>A custom node can represent a part of the parsed language, a type of node present in an AST.